stab_controller_multi_lqr_node.cpp

```diff
@@ -0,0 +1,416 @@
+For read in of the k matrix array the matrices should be stored in "/resources"
+as a .csv file and named as "k_matrix_0", "k_matrix_1", ...
+For read in of the imu_offset scaler array the vectors should be stored
+in "/resources" as a .csv file and named as "imu_offset_0", "imu_offset_1", ...
+--------------------------------------------------------------------------------
+TODO: is double necessary in matrices or would float be enough?
+--------------------------------------------------------------------------------
+*/
+
+//Include header, ros and other directories
+#include <ros/ros.h>
+#include "stab_controller_multi_lqr_node.h"
+
+//Set namespace
+namespace ascento {
+
+  //Constructor / Destructor
+  StabControllerMultiLqrNode::StabControllerMultiLqrNode(ros::NodeHandle& nh) : maxon_left(1,"USB1", "EPOS4"), maxon_right(2,"USB0", "EPOS4"), nh(nh) {
+    ROS_INFO("stab_controller_multi_lqr node has started.");
+
+    //Initialise class members
+    Initialise();
+
+    //Initialise Threads for reading position/velocity and writing torques to maxon motors
+    std::thread ReadLeft(&StabControllerMultiLqrNode::ReadLeft,this);
+    std::thread ReadRight(&StabControllerMultiLqrNode::ReadRight,this);
+    std::thread WriteLeft(&StabControllerMultiLqrNode::WriteLeft,this);
+    std::thread WriteRight(&StabControllerMultiLqrNode::WriteRight,this);
+
+    // Start rated loop
+    RateController();
+
+  }
+  StabControllerMultiLqrNode::~StabControllerMultiLqrNode() {
+    maxon_left.MoveWithTorque(0);
+    maxon_right.MoveWithTorque(0);
+  }
+
+  //Setup methods
+  void StabControllerMultiLqrNode::ResetStates() {
+    encoder_offset_left = maxon_left.GetPosition()*2*M_PI/encoder_quadcounts;
+    encoder_offset_right = maxon_right.GetPosition()*2*M_PI/encoder_quadcounts;
+  }
+  void StabControllerMultiLqrNode::Initialise() {
+    // Read parameters from server
+    if (!ReadParameters()) {
+      ROS_ERROR("Could not read parameters.");
+      ros::requestShutdown();
+    }
+
+    InitSubPub();
+    InitSrv();
+
+    // Setup Controller
+    ResetStates();
+    Eigen::VectorXd u_sat = Eigen::MatrixXd(NO_OF_INPUTS,1);
+    u_sat << saturation_ec90;
+    controller.SetUSat(u_sat);
+    input_delays_left.setZero(NO_OF_DELAYS, 1);
+    input_delays_right.setZero(NO_OF_DELAYS, 1);
+    input_left.setZero(NO_OF_INPUTS, 1);
+    input_right.setZero(NO_OF_INPUTS, 1);
+    meas_left.setZero(NO_OF_MEASUREMENTS, 1);
+    meas_right.setZero(NO_OF_MEASUREMENTS, 1);
+
+
+    // Read in K matrices
+    ReadInMatrixArray("/home/ascento/catkin_ws/src/ascento/Matlab/DynMod/fixed_drive/ros_compatible/k_matrix_");
+    // Read in Imu offsets
+    ReadInDoubleArray("/home/ascento/catkin_ws/src/ascento/Matlab/DynMod/fixed_drive/ros_compatible/imu_offset");
+  }
+  void StabControllerMultiLqrNode::InitSubPub() {
+    //Initialise subscribers
+    imu_sub = nh.subscribe("/est_states/imu", 1, &StabControllerMultiLqrNode::SaveImuCallback, this);
```

Comment:
It is good practice to make class variables recognizably different from function scope variables, for example by appending a trailing _ to their names